File: leetcode/0050.go
package leetcode

func myPow(x float64, n int) float64 {
	q_pow := func(x float64, n int) (ans float64) {
		ans = 1.0
		for n > 0 {
			if n&0x1 == 1 {
				ans *= x
			}
			x *= x
			n >>= 1
		}
		return
	}
	if n == 0 {
		return 1.0
	} else if n < 0 {
		return 1.0 / q_pow(x, -n)
	}
	return q_pow(x, n)

}

File: leetcode/0050.cpp
class Solution {
public:

    double myPow(double x, int n) {
        if (!n)
            return 1.0;
        auto q_pow = [&x](long long n){
            auto ans = 1.0;
            while(n){
                if (n&1)
                    ans *= x;
                x *=x;
                n >>= 1;
            }
            return ans;
        };
        long N = n;
        if (n < 0){
            return 1.0/q_pow(-N);
        }
        return q_pow(N);
    }
};